File: Tyche.Backend/Pipel.Tyche.Pool.Api/Pipel.Tyche.Pool.Api/FindPoolsGamblersUseCase.fs
module Pipel.Tyche.Pool.Api.FindPoolsGamblersUseCase

open Pipel.Core
open Pipel.Data
open Pipel.Tyche.Pool.Api
open Pipel.Tyche.Pool.Domain
open Pipel.Tyche.Pool.Domain.UseCases
open Pipel.Type

let execute
    (poolId: string)
    (filter: string)
    (next: string)
    (findPoolsGamblersUseCase: IFindPoolsGamblersUseCase)
    (mapFromDomainToViewFunc: MapFunc<PoolGambler, PoolGamblerResponse>)
    =
    async {
        let! page =
            findPoolsGamblersUseCase.AsyncExecute(
                { PoolPK.PoolId = Ulid.From poolId },
                filter |> String.tryFrom,
                next |> String.tryFrom
            )

        return page |> CursorPage.map mapFromDomainToViewFunc
    }

File: Tyche.Android/app/src/main/java/com/pipel/tyche/poolGambler/useCase/FindPoolsGamblersUseCase.kt
package com.pipel.tyche.poolGambler.useCase

import com.pipel.core.CursorPage
import com.pipel.tyche.poolGambler.data.PoolGamblerRepository
import com.pipel.tyche.poolGambler.domain.PoolGambler
import com.pipel.tyche.poolGambler.mapper.PoolGamblerMapper
import javax.inject.Inject

interface FindPoolsGamblersUseCase {

    suspend fun execute(
        poolId: String,
        nextToken: String?,
        filter: String?
    ): CursorPage<PoolGambler>

}

class DefaultFindPoolsGamblersUseCase @Inject constructor(private val poolGamblerRepository: PoolGamblerRepository) :
    FindPoolsGamblersUseCase {

    override suspend fun execute(
        poolId: String,
        nextToken: String?,
        filter: String?
    ): CursorPage<PoolGambler> {
        return poolGamblerRepository.getPoolsGamblers(poolId, nextToken, filter)
            .map(PoolGamblerMapper::mapFromDataToDomain)
    }

}